beluga_ros/include/beluga_ros/point_cloud.hpp

```diff
@@ -0,0 +1,102 @@
+// Copyright 2024 Ekumen, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef BELUGA_ROS_POINT_CLOUD_HPP
+#define BELUGA_ROS_POINT_CLOUD_HPP
+
+#include <range/v3/view/iota.hpp>
+
+#include <beluga/sensor/data/point_cloud.hpp>
+#include <beluga/views/take_evenly.hpp>
+#include <beluga_ros/messages.hpp>
+
+#include <sophus/se3.hpp>
+
+#include <Eigen/Dense>
+
+#include "beluga/eigen_compatibility.hpp"
+
+/**
+ * \file
+ * \brief Implementation of `sensor_msgs/PointCloud2` wrapper type for messages with with memory-aligned strides.
+ *
+ * \details
+ * The stride calculation ensures that the memory layout of the point cloud data is aligned with the size of the data
+ * type used for iteration (`iteratorType`). To maintain proper memory alignment, the stride must satisfy the condition:
+ *
+ * `cloud_->point_step % sizeof(iteratorType) == 0`
+ *
+ * This condition guarantees that `point_step` is a multiple of the size of the `iteratorType`, ensuring efficient
+ * access to each point in the cloud.
+ */
+
+namespace beluga_ros {
+
+/// Thin wrapper type for 3D `sensor_msgs/PointCloud2` messages.
+/// Assumes an XYZ... type message.
+/// All datafields must be the same type (float or double).
+template <uint8_t T>
+class PointCloud3 : public beluga::BasePointCloud<PointCloud3<T>> {
+ public:
+  /// PointCloud data fields type
+  using Scalar = typename sensor_msgs::pointFieldTypeAsType<T>::type;
+
+  /// Check type is float or double
+  static_assert(
+      std::is_same<Scalar, float>::value || std::is_same<Scalar, double>::value,
+      "Pointcloud3 only supports float or double datatype");
```

Comment:
Mmmm Since most of the lidar use float or double for the XY&Z data I thought it would be convenient to check that the message wrapper actually uses that data type. Although, I think the wrapper can work with any type